### ISSUE TITLE
Build Header lists specfic commit but worker checks out latest merge

### ISSUE THREAD

**Issue by nnutter** [travis-ci/travis-ci#3387]
We noticed today that a [build/job](https://travis-ci.org/genome/TestDbServer/builds/54615621) we restarted listed itself as testing a specific commit when in fact it was testing another.  The Build Header says, "Commit 332a7f7", but the logs show it just checking out the latest merge commit for the pull request,

```
$ git fetch origin +refs/pull/68/merge:
remote: Counting objects: 11, done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 11 (delta 6), reused 5 (delta 1), pack-reused 0
Unpacking objects: 100% (11/11), done.
From git://github.com/genome/TestDbServer
* branch refs/pull/68/merge -> FETCH_HEAD
$ git checkout -qf FETCH_HEAD
```

Before noticing this discrepancy I was lead to think that we had an intermittent failure but really it was due to a code change.  It seems like either restarting a build should be sure to check out the same revision or it should at least update the Commit field in the Build Header to not be incorrect.

P.S. Thanks for making a great service!

**Comment by rkh** [comment]
Thanks for the report. We use indeed the latest merge, but it's the only one available. The alternative would be to error out.

Relevant discussion: https://github.com/travis-ci/travis-build/pull/26

**Comment by joshk** [comment]
@rkh or we could cancel running builds if new PR builds come in?

**Comment by nnutter** [comment]
It seems like the Commit in the Build Header should at least reflect the actual commit that is being used for that Build.  That's a little tricky because as it works now it seems like Travis CI couldn't actually know that until after the Build starts.  So it might be best just to remove the Commit from the Build Header entirely.  (Just brainstorming.)

However, I think a bigger issue is that the status being returned to GitHub seems to be based on that Commit.  In my example Travis CI updated the status on 7dc475d because that was the commit that was merged in on the original merge commit, 332a7f7, but when the Build ran it was based on a later merge due to a later commit in the pull request.  So Travis CI ran tests on Y but reported it on X.

It seems like the best thing would be for GitHub to make non-ephemeral refs for the merges, e.g. `refs/pull/68/merge-09e17c1-7dc475d`, so that you guys could actually avoid both the race condition and allow for restarts.

**Comment by Cyberboss** [comment]
If you want to make sure the pull request is built against the latest commit of its base branch you have to compare the creation time of the refs/pull/{number}/merge commit against the creation time of the build. Needless to say, this is not ideal.
